### ISSUE TITLE
Mark as native function after lockdown

### ISSUE THREAD

**Issue by Jack-Works** [endojs/endo#1199]
I'm working on my eval-less Compartment and I want to mark my function as [native code]. Since it's already running in the SES, I cannot change Object.prototype.toString, so it will be good if I can mark my class/functions as `native code`.

Possible solution:

- Expose a global function
- Implement <https://github.com/tc39/proposal-function-implementation-hiding>

**Comment by erights** [comment]
The current shim behavior --- where an extensible weakmap side table is consulted by the replacement `Function.prototype.toString` --- is unlikely to be something we'd want to propose. Rather, the hardened js proposal implies that the post-lockdown environment would by default contain only native functions. So I'm reluctant to expose a global function, as that would be more non-proposed API that code may come to depend on. 

Do you have a practical strategy for shimming the implementation hiding proposal? If so, that may indeed be the better direction to look.

**Comment by Jack-Works** [comment]
Call `.toString()` on the function, if it contains `"hide source"` or `'hide source'`, then return `[native code]`. But that requires `toString()` on all functions so I think extensible WeakMap is better since it does not trigger arbitrary user code.

It can be an undocumented API with a special name, so only shim authors will use it, like `lockdown.__shim__mark__as__native__`

**Comment by Jack-Works** [comment]
![image](https://user-images.githubusercontent.com/5390719/176096006-76f3b467-364c-4ef4-9996-3e9dcce29719.png)

I found sentry is doing the same thing and I'm figuring out how to use sentry with lockdown

**Comment by Jack-Works** [comment]
another idea comes up to my mind. If we allows `Function.prototype.toString()` to be overwritten after lockdown, it allows those libraries to patch them as "native function"

**Comment by erights** [comment]
"after lockdown"? How could that possibly be safe?

Do you mean, during the (hypothetical) vetted shim phase, that happens after the `repairIntrinsics` phase of lockdown, but before the `hardenIntrinsics` phase? This hypothetical phase is when we hope to run so-called vetted shims, that can indeed do further patching of the environment before the intrinsics get hardened. Unfortunately, other than split lockdown into those two phases, we have not done much yet to support running vetted shims.

**Comment by Jack-Works** [comment]
Since we don't have a `repairIntrinsics` phase... But marking function as native is a thing that many implementation does.

**Comment by Jack-Works** [comment]
Like what we did the tolerate to Array.prototype.sort.prototype

**Comment by erights** [comment]
What I don't understand is the suggestion to be able to do it after lockdown. This would imply that lockdown would leave `Function.prototype.toString` in a mutable state (writable or configurable or both). It would them be a communications channel, preventing compartments from being isolated from each other,

**Comment by Jack-Works** [comment]
Yes, this will be unsafe. But any method I can think of is either unsafe or will trigger user code.

**Comment by erights** [comment]
See also https://github.com/endojs/endo/issues/1228

**Comment by erights** [comment]
@Jack-Works now that #1228 is fixed, do you still need this feature?

**Comment by Jack-Works** [comment]
> @Jack-Works now that #1228 is fixed, do you still need this feature?

Yes, they're two different things. This issue is for third-party polyfill to hide their sources.
